Add support for windows desktop services proxying different desktops

lxea opened the pull request:
This changes the key for windows desktops to be formatted like `windowsDesktop/$host_id/$desktop_name` as a result all the Get/Delete API for desktops was updated.

Also adds a `GetWindowsDesktopByName` endpoint which provides all desktops with the same desktop name but different host IDs.

Commit:
Add support for windows desktop services proxying different desktops

zmb3 commented on lib/web/desktop.go:
```diff
@@ -103,17 +102,36 @@ func createDesktopConnection(
+			ConnType: types.WindowsDesktopTunnel,
+			ServerID: service.GetName() + "." + ctx.parent.clusterName,
+		})
+		if err != nil {
```
Don't forget, if err is nil then you connected successfully and you don't need to retry, you can break out of the for loop.

nklaassen commented on lib/web/desktop.go:
```diff
@@ -103,17 +102,35 @@ func createDesktopConnection(
+	for _, desktop := range winDesktops {
+		for _, service := range winServices {
+			if desktop.GetHostID() != service.GetName() {
+				continue outer
```
why outer, no other service will match this desktop?

zmb3 replied:
yeah, I don't think this check is necessary.

nklaassen commented on lib/web/desktop.go:
```diff
@@ -103,17 +102,35 @@ func createDesktopConnection(
-		ConnType: types.WindowsDesktopTunnel,
-		ServerID: service.GetName() + "." + ctx.parent.clusterName,
-	})
+	winDesktops, err := ctx.unsafeCachedAuthClient.GetWindowsDesktopsByName(r.Context(), desktopName)
```
check `err`?

nklaassen commented on lib/web/desktop.go:
```diff
@@ -103,17 +102,35 @@ func createDesktopConnection(
 	//
```
In the comment above
> Pick a random Windows desktop service as our gateway

is this still random? should it be?

lxea replied:
Ah it should be, changed to shuffle it, I believe the database server selection is also shuffled.

nklaassen commented on api/client/proto/authservice.proto:
```diff
@@ -1160,16 +1160,27 @@ message GetWindowsDesktopsResponse {
 }
 
+// GetWindowsDesktopsByNameRequest is a request for a collection of
+// Windows desktops by name host.
```
> by name host.

host name?

zmb3 commented on api/client/proto/authservice.proto:
```diff
@@ -1160,16 +1160,27 @@ message GetWindowsDesktopsResponse {
 message DeleteWindowsDesktopRequest {
     // Name is the name of the Windows desktop host.
     string Name = 1 [ (gogoproto.jsontag) = "name" ];
+    // HostID is the ID of the Windows desktop service.
```
```suggestion
    // HostID is the ID of the Windows Desktop Service reporting the desktop.
```

zmb3 commented on api/client/proto/authservice.proto:
```diff
@@ -1798,6 +1809,9 @@ service AuthService {
     // GetWindowsDesktops returns all registered Windows desktop hosts.
     rpc GetWindowsDesktops(google.protobuf.Empty) returns (GetWindowsDesktopsResponse);
+    // GetWindowsDesktopsByName returns all registered Windows desktop hosts matching name.
+    rpc GetWindowsDesktopsByName(GetWindowsDesktopsByNameRequest)
```
Do we need this new rpc?

Can we just use `GetWindowsDesktop()` where we set name and leave `HostID` as empty string?

lxea replied:
I think its a little clearer/makes calls a little nicer, to have separate rpcs as `GetWindowsDesktop` always returnes a single item and `GetWindowsDesktopsByName` always expects a list.

Do other rpcs follow a similar pattern of "" acting like a glob key or something?

zmb3 replied:
It's not so much of a glob as it is a filter.

- `rpc ListNodes(ListNodesRequest)` lists nodes that meet the criteria of the request.
- `rpc GetAccessRequests(AccessRequestFilter)` gets access requests filtered by the paramters in the request, and unset means no filter, just like I proposed here
- `rpc GetPluginData(PluginDataFilter)` behaves similarly.

lxea replied:
Ah thats pretty cool, maybe it would be worth changing it so instead of having multiple rpc endpoints for getting desktops, there could just be GetWindowsDesktops that has a filter type for requests instead.

zmb3 commented on lib/web/desktop.go:
```diff
@@ -103,17 +102,35 @@ func createDesktopConnection(
+	winDesktops, err := ctx.unsafeCachedAuthClient.GetWindowsDesktopsByName(r.Context(), desktopName)
+	var serviceCon net.Conn
+outer:
+	for _, desktop := range winDesktops {
```
I would shuffle `winDesktops` first before looping through them. Otherwise, a bunch of attempts to connect to the same desktop may all go through the same Windows Desktop Service, rather than sharing the load equally.

zmb3 commented on lib/web/servers.go:
```diff
@@ -90,10 +90,14 @@ func (h *Handler) getDesktopHandle(w http.ResponseWriter, r *http.Request, p htt
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
+	windowsDesktops = types.DeduplicateDesktops(windowsDesktops)
```
In this case, I don't think you need to deduplicate since they're all the same from the UI's perspective.

As long as you got at least 1 desktop, you should be able to just

```
return ui.MakeDesktop(windowsDesktops[0]), nil
```

zmb3 commented on tool/tctl/common/resource_command.go:
```diff
@@ -740,9 +740,23 @@ func (rc *ResourceCommand) Delete(client auth.ClientI) (err error) {
+			}
+		}
+		if !deleted {
+			return trace.NotFound("desktop %q not found", rc.ref.Name)
```
1. This could fail for reasons other than not found, so I wouldn't assume it's a not found error.
2. I would probably return a `trace.Aggregate(...)` of all the delete errors.

zmb3 reviewed (commented):
Looking good. Left a few comments. Let me know when you've addressed them and we can test this out together!

zmb3 commented on lib/web/desktop.go:
```diff
@@ -103,17 +103,41 @@ func createDesktopConnection(
-		To:       &utils.NetAddr{AddrNetwork: "tcp", Addr: service.GetAddr()},
-		ConnType: types.WindowsDesktopTunnel,
-		ServerID: service.GetName() + "." + ctx.parent.clusterName,
+	rand.Shuffle(len(winServices), func(i, j int) {
```
You're still fetching _all_ Windows Desktop Services, then shuffling them, and then searching through them until you find one that works for this desktop.

If there are thousands of Windows Desktop Services and only 2-3 of them are actually eligible candidates, this is wasteful.

What if you did something like this (pseudocode) instead?

```
desktops, _ := GetWindowsDesktopsByName(desktopName)

var eligibleServiceIDs []string
for _, desktop := range desktops {
    eligibleServiceIDs = append(eligibleServiceIDs, desktop.GetHostID())
}
rand.Shuffle(eligibleServiceIDs, ...)

// try each eligible service, one at a time
// (this will likely be a much smaller list than your original for loop)
for _, id := eligibleServiceIDs {
    wds, _ := GetWindowsDesktopService(id) // this would be a new RPC
    serviceCon, err = site.DialTCP(...)
}
```

zmb3 commented on lib/web/servers.go:
```diff
@@ -90,10 +90,13 @@ func (h *Handler) getDesktopHandle(w http.ResponseWriter, r *http.Request, p htt
 		return nil, trace.Wrap(err)
 	}
+	if len(windowsDesktops) == 0 {
+		return nil, trace.Errorf("Expected one desktop, got %d", len(windowsDesktops))
```
- You didn't expect one desktop, you expected at least one desktop.
- In this case, I think `trace.NotFound` is appropriate.

zmb3 commented on tool/tctl/common/resource_command.go:
```diff
@@ -740,9 +740,28 @@ func (rc *ResourceCommand) Delete(client auth.ClientI) (err error) {
+			errs = append(errs,
+				trace.Errorf("failed to delete desktop %q", rc.ref.Name))
+		}
+		if len(errs) != 0 {
```
You don't actually need the `len` check. `NewAggregate` will do it for you and correctly return `nil`.

Commit:
Add filter to GetWindowsDesktops, remove GetWindowsDesktop and GetWindowsDesktopByName

jakule commented on lib/auth/auth_with_roles.go:
```diff
@@ -3834,18 +3831,19 @@ func (a *ServerWithRoles) UpsertWindowsDesktop(ctx context.Context, s types.Wind
 		return trace.Wrap(err)
 	}
-	if err := a.checkAccessToWindowsDesktop(desktop); err != nil {
+	if err := a.checkAccessToWindowsDesktop(desktop[0]); err != nil {
```
Can we check if the slice is empty first?

jakule commented on lib/auth/auth_with_roles.go:
```diff
@@ -3818,9 +3814,10 @@ func (a *ServerWithRoles) UpsertWindowsDesktop(ctx context.Context, s types.Wind
+		types.WindowsDesktopFilter{HostID: s.GetHostID(), Name: s.GetName()})
 	if err == nil {
-		if err := a.checkAccessToWindowsDesktop(existing); err != nil {
+		if err := a.checkAccessToWindowsDesktop(existing[0]); err != nil {
```
Can we check if the slice is not empty? This potentially can panic.

zmb3 commented on api/client/proto/authservice.proto:
```diff
@@ -1151,16 +1163,12 @@ message GetWindowsDesktopsResponse {
 message DeleteWindowsDesktopRequest {
     // Name is the name of the Windows desktop host.
     string Name = 1 [ (gogoproto.jsontag) = "name" ];
+    // HostID is the ID of the Windows Desktop Service reporting the desktop.
```
I would add to the comment. Something like:

```
// HostID is the ID of the Windows Desktop Service reporting the desktop.
// If not specified, all desktops matching Name will be deleted.
```

zmb3 commented on api/client/proto/authservice.proto:
```diff
@@ -1151,16 +1163,12 @@ message GetWindowsDesktopsResponse {
     // Name is the name of the Windows desktop host.
     string Name = 1 [ (gogoproto.jsontag) = "name" ];
+    // HostID is the ID of the Windows Desktop Service reporting the desktop.
+    string HostID = 2 [ (gogoproto.jsontag) = "HostID" ];
```
```suggestion
    string HostID = 2 [ (gogoproto.jsontag) = "host_id" ];
```

zmb3 commented on lib/web/desktop.go:
```diff
@@ -87,34 +88,57 @@ func createDesktopConnection(
+			return trace.WrapWithMessage(err,
+				"error connecting to windows_desktop_service at %q: %v", service.GetAddr(), err)
+		}
+		log.Errorf("Error connecting to service %q, trying another.", service.GetAddr())
```
I think this should be a warning or even info. It's only an error if we can't connect to any of them, and you've already got that case covered.

zmb3 commented on api/client/proto/authservice.proto:
```diff
@@ -1151,16 +1163,13 @@ message GetWindowsDesktopsResponse {
-    string Name = 1 [ (gogoproto.jsontag) = "name" ];
-}
-
 // DeleteWindowsDesktopRequest is a request to delete a Windows desktop host.
```
I would update the godoc here, since this can now delete one or more hosts.

zmb3 commented on api/types/types.proto:
```diff
@@ -2689,6 +2689,11 @@ message WindowsDesktopServiceSpecV3 {
     string TeleportVersion = 2 [ (gogoproto.jsontag) = "teleport_version" ];
 }
 
+message WindowsDesktopFilter {
```
Any particular reason this is in types.proto instead of authservice.proto?

lxea replied:
WebSessionFilter and SemaphoreFilter were all in types.proto so it seemed most consistent to keep the filter types in there.

zmb3 commented on api/types/types.proto:
```diff
@@ -2759,6 +2764,8 @@ message WindowsDesktopSpecV3 {
     string Addr = 1 [ (gogoproto.jsontag) = "addr" ];
     // Domain is the ActiveDirectory domain that this host belongs to.
     string Domain = 2 [ (gogoproto.jsontag) = "domain" ];
+    // HostID is the ID of the host the desktop server is running on.
```
```suggestion
    // HostID is the ID of the Windows Desktop Service proxying the desktop.
```

zmb3 commented on lib/services/local/desktops.go:
```diff
@@ -139,15 +131,41 @@ func (s *WindowsDesktopService) UpsertWindowsDesktop(ctx context.Context, deskto
-	if err != nil {
-		if trace.IsNotFound(err) {
-			return trace.NotFound("windows desktop %q doesn't exist", name)
+func (s *WindowsDesktopService) DeleteWindowsDesktop(ctx context.Context, hostID, name string) error {
```
What happens if `name` is empty too? Would it delete all?

I wonder if we should make that an error, since there's a separate RPC to delete all.

zmb3 commented on lib/web/desktop.go:
```diff
@@ -87,34 +88,57 @@ func createDesktopConnection(
-		return trace.WrapWithMessage(err, "failed to connect to windows_desktop_service at %q: %v", service.GetAddr(), err)
+		return trace.Errorf("failed to connect to any windows_desktop_service: %v", err)
 	}
 	defer serviceCon.Close()
```
When I run this, I get a panic here. Looks like we're assuming that `serviceCon` cannot be nil if we didn't get an error.

I would add a check after the error check.

```
if serviceCon == nil {
    return trace.Errorf("...")
}
```

zmb3 reviewed (changes requested):
This also fails on clusters where you've upgraded from a version prior to this change existing:

```
6T14:18:55-07:00 ERRO [WINDOWS_D] "desktop reconciliation failed: failed to update windows_desktop WIN-I44F9TN11M3-zacmanual-example-com\n\tno windows desktops with HostID 316a3ffa-23e6-4d85-92a1-5e44754f8189 and Name WIN-I44F9TN11M3-zacmanual-example-com" desktop/discovery.go:101
```

zmb3 commented on lib/auth/auth_with_roles.go:
```diff
@@ -3902,9 +3903,14 @@ func (a *ServerWithRoles) UpsertWindowsDesktop(ctx context.Context, s types.Wind
+	existing, err := a.authServer.GetWindowsDesktops(ctx,
+		types.WindowsDesktopFilter{HostID: s.GetHostID(), Name: s.GetName()})
+	if len(existing) == 0 {
+		return trace.NotFound("no windows desktops with HostID %s and Name %s",
```
Ah, here's the bug. This is Upsert, so it should update if already exists, and create if it doesn't already exist.

Here we're returning an error if it doesn't exist, so it will never be created.

zmb3 committed:
Cache cleanup

zmb3 committed:
Fix cache deletes for Windows desktops

For deletes, the cache only gets the backend key, not the entire
resource. Do what database access does, which is to extract the
host ID from the path, and stuff it in the description field of
the resource header.

zmb3 committed:
Godoc cleanup

zmb3 commented:
@jakule @nklaassen I'd like to get this merged today. Can you please take another pass?

nklaassen commented on build.assets/version-check/main.go:
```diff
@@ -0,0 +1,151 @@
+/*
```
is this meant to be in this PR? Could it be pulled out into its own?

jakule replied:
I was about to ask about the same.

nklaassen commented on lib/web/desktop.go:
```diff
@@ -87,34 +88,65 @@ func createDesktopConnection(
 	})
+
+	var serviceCon net.Conn
+	for _, id := range validServiceIDs {
+		service, err := ctx.clt.GetWindowsDesktopService(context.Background(), id)
+		if err != nil {
+			log.Errorf("Error finding service with id %s", id)
+			continue
+		}
+		log = log.WithField("windows-service-uuid", service.GetName())
+		log = log.WithField("windows-service-addr", service.GetAddr())
+		serviceCon, err = site.DialTCP(reversetunnel.DialParams{
+			From:     &utils.NetAddr{AddrNetwork: "tcp", Addr: r.RemoteAddr},
+			To:       &utils.NetAddr{AddrNetwork: "tcp", Addr: service.GetAddr()},
+			ConnType: types.WindowsDesktopTunnel,
+			ServerID: service.GetName() + "." + ctx.parent.clusterName,
+		})
+		if err == nil {
+			break
+		}
+		if !trace.IsConnectionProblem(err) {
+			return trace.WrapWithMessage(err,
+				"error connecting to windows_desktop_service at %q: %v", service.GetAddr(), err)
+		}
+		log.Infof("Error connecting to service %q, trying another.", service.GetAddr())
+	}
 	if err != nil {
-		return trace.WrapWithMessage(err, "failed to connect to windows_desktop_service at %q: %v", service.GetAddr(), err)
+		return trace.Errorf("failed to connect to any windows_desktop_service: %v", err)
+	}
+	if serviceCon == nil {
+		return trace.Errorf("failed to connect to any windows_desktop_service: connection unexpectedly nil")
 	}
```
nit: I like to see loops like this pulled into a function so you can just `return` rather than `break`ing out of the loop, and you don't even need the mutable `serviceCon` variable that you have to check for nil after the loop, but I won't block the PR for this

nklaassen reviewed (commented):
LGTM except, is the `version-check` stuff meant to be in this PR? I don't see where it is used

zmb3 committed:
Fix lint

zmb3 committed:
Address review comments

zmb3 committed:
Send error message if no desktop found

zmb3 committed:
Revert to x/net/websocket

This got converted to gorilla/websocket as part of moderated sessions.
We'll do a more intentional conversion post-release.

zmb3 committed:
fix lint
